File: bench/bench_soundml.py
"""Librosa cross-reference benchmarks for SoundML.

The Python twin of ``bench_soundml.ml``: every SoundML case with a librosa (or
scipy) reference implementation gets a row here under the same group/name
scheme, so the two tables read side by side. Window generation seeds the
suite; STFT and mel rows land together with their OCaml counterparts.

Run it from the repository root:

    python bench/bench_soundml.py
"""

from __future__ import annotations

import time
from typing import Any, Callable, List, Tuple

import librosa
import numpy as np

N_WINDOW = 2048

# The SoundML window specs and their librosa/scipy spellings. Parametrised
# specs use representative parameters; keep them in sync with the OCaml suite.
WINDOW_SPECS: List[Tuple[str, Any]] = [
    ("hann", "hann"),
    ("hamming", "hamming"),
    ("blackman", "blackman"),
    ("blackman_harris", "blackmanharris"),
    ("nuttall", "nuttall"),
    ("bartlett", "bartlett"),
    ("kaiser 8.6", ("kaiser", 8.6)),
    ("gaussian 256", ("gaussian", 256.0)),
    ("tukey 0.5", ("tukey", 0.5)),
    ("flat_top", "flattop"),
    ("rectangular", "boxcar"),
]


def bench(name: str, fn: Callable[[], Any]) -> Tuple[str, Callable[[], Any]]:
    return (name, fn)


def build_benchmarks() -> List[Tuple[str, Callable[[], Any]]]:
    benches: List[Tuple[str, Callable[[], Any]]] = []

    for name, spec in WINDOW_SPECS:
        def make(spec: Any) -> Callable[[], Any]:
            def run() -> None:
                w = librosa.filters.get_window(spec, N_WINDOW, fftbins=True)
                float(w.sum())

            return run

        benches.append(bench(f"window/{name} {N_WINDOW} (librosa)", make(spec)))

    return benches


def measure(
    fn: Callable[[], Any],
    *,
    warmup: int = 3,
    batches: int = 7,
    min_batch_seconds: float = 0.02,
) -> float:
    """Median wall-clock seconds per call, over ``batches`` timed batches.

    The batch size is calibrated by doubling until one batch takes at least
    ``min_batch_seconds``, so per-call durations stay resolvable for
    microsecond-scale cases.
    """
    for _ in range(warmup):
        fn()
    calls = 1
    while True:
        start = time.perf_counter()
        for _ in range(calls):
            fn()
        elapsed = time.perf_counter() - start
        if elapsed >= min_batch_seconds:
            break
        calls *= 2
    samples = []
    for _ in range(batches):
        start = time.perf_counter()
        for _ in range(calls):
            fn()
        samples.append((time.perf_counter() - start) / calls)
    return float(np.median(samples))


def format_seconds(seconds: float) -> str:
    if seconds < 1e-6:
        return f"{seconds * 1e9:.2f}ns"
    if seconds < 1e-3:
        return f"{seconds * 1e6:.2f}us"
    if seconds < 1.0:
        return f"{seconds * 1e3:.2f}ms"
    return f"{seconds:.2f}s"


def main() -> None:
    benchmarks = build_benchmarks()
    width = max(len(name) for name, _ in benchmarks)
    print(f"{'Name':<{width}}  Wall/Run")
    print("-" * (width + 10))
    for name, fn in benchmarks:
        wall = measure(fn)
        print(f"{name:<{width}}  {format_seconds(wall):>8}")


if __name__ == "__main__":
    main()

File: bench/bench_soundml.ml
(* SoundML benchmarks.

   The window group times [Window.make] for every spec at n = 2048, one row per
   spec, mirroring the librosa rows of [bench_soundml.py] one to one.

   The pipeline group prices the abstraction: [Pipeline.run] of a three-stage
   stateless toy chain over a one-million-sample float32 tensor, next to the
   hand-written sequence of the same three Nx calls. The two rows must stay
   within a few percent of each other; the committed baseline plus the
   suite-level budgets below turn any drift into a failed build.

   The stft and mel groups time the offline analysis paths end to end —
   [Stft.power_spectrum] and [mel_spectrogram] over 30 s of mono audio at 22.05
   kHz, fft 2048 and hop 512, in both float32 and float64 — one row per dtype,
   mirroring the librosa rows of [bench_soundml.py] one to one. *)

let n_window = 2048

(* One (name, spec) pair per Window.t constructor; parametrised specs use the
   same representative parameters as bench_soundml.py. *)
let window_specs =
  [ ("hann", Soundml.Window.Hann)
  ; ("hamming", Soundml.Window.Hamming)
  ; ("blackman", Soundml.Window.Blackman)
  ; ("blackman_harris", Soundml.Window.Blackman_harris)
  ; ("nuttall", Soundml.Window.Nuttall)
  ; ("bartlett", Soundml.Window.Bartlett)
  ; ("kaiser 8.6", Soundml.Window.Kaiser 8.6)
  ; ("gaussian 256", Soundml.Window.Gaussian 256.)
  ; ("tukey 0.5", Soundml.Window.Tukey 0.5)
  ; ("flat_top", Soundml.Window.Flat_top)
  ; ("rectangular", Soundml.Window.Rectangular) ]

let window_benchmarks () =
  List.map
    (fun (name, spec) ->
      Thumper.bench (Printf.sprintf "%s %d" name n_window) (fun () ->
          Soundml.Window.make Nx.float64 spec n_window ) )
    window_specs

(* 30 s of mono audio at librosa's default rate, analysed at librosa's default
   fft size and hop; the mel projection uses librosa's default 128 bands. *)
let sample_rate = 22050

let n_audio = 30 * sample_rate

let stft_config = Soundml.Stft.Config.create ~fft_size:2048 ~hop:512 ()

let mel_config =
  Soundml.Mel.Config.create ~n_mels:128 ~sample_rate ~fft_size:2048 ()

let stft_benchmarks () =
  let x32 = Nx.rand Nx.float32 [|n_audio|] in
  let x64 = Nx.rand Nx.float64 [|n_audio|] in
  [ Thumper.bench "power_spectrum 30s f32 fft2048 hop512" (fun () ->
        Soundml.Stft.power_spectrum stft_config x32 )
  ; Thumper.bench "power_spectrum 30s f64 fft2048 hop512" (fun () ->
        Soundml.Stft.power_spectrum stft_config x64 ) ]

let mel_benchmarks () =
  let x32 = Nx.rand Nx.float32 [|n_audio|] in
  let x64 = Nx.rand Nx.float64 [|n_audio|] in
  [ Thumper.bench "mel_spectrogram 30s f32 fft2048 hop512" (fun () ->
        Soundml.mel_spectrogram stft_config mel_config x32 )
  ; Thumper.bench "mel_spectrogram 30s f64 fft2048 hop512" (fun () ->
        Soundml.mel_spectrogram stft_config mel_config x64 ) ]

let n = 1_000_000

let source =
  Soundml.Pipeline.Format.audio Nx.float32 ~sample_rate:44100 ~channels:1

(* The toy chain: gain, bias, rectify — three stateless stages whose
   hand-written equivalent is exactly three Nx calls. *)
let chain =
  let open Soundml.Pipeline in
  stateless (fun t -> Nx.mul_s t 0.5)
  >> stateless (fun t -> Nx.add_s t 0.1)
  >> stateless Nx.abs

let direct t = Nx.abs (Nx.add_s (Nx.mul_s t 0.5) 0.1)

let pipeline_benchmarks () =
  let x = Nx.rand Nx.float32 [|n|] in
  [ Thumper.bench "run gain>>bias>>abs 1M" (fun () ->
        Soundml.Pipeline.run ~source chain x )
  ; Thumper.bench "direct gain+bias+abs 1M" (fun () -> direct x) ]

let () =
  Nx.Rng.run ~seed:42
  @@ fun () ->
  Thumper.run "soundml"
    ~budgets:
      [ Thumper.Budget.no_slower_than ~metric:Thumper.Metric.wall_time 0.05
      ; Thumper.Budget.no_more_alloc_than 0.01 ]
    [ Thumper.group "window" (window_benchmarks ())
    ; Thumper.group "pipeline" (pipeline_benchmarks ())
    ; Thumper.group "stft" (stft_benchmarks ())
    ; Thumper.group "mel" (mel_benchmarks ()) ]
